### PR TITLE
Make nsIUserInfo component optional with --disable-userinfo at build time

### LINKED ISSUE #795: Make nsUserInfo component optional at build time.
While investigating some things to speedup the startup, I bumped into nsUserInfo, an old mostly abandoned codebase that [was going to be used for TB22](https://bugzilla.mozilla.org/show_bug.cgi?id=29593) to auto-populate fields. It appears to not be used anywhere at the moment, and isn't working in the latest TB.

As an aside, this module scrapes full username, e-mail, and hostname and makes it available to all applications using the toolkit. So I suppose it is a privacy improvement. Test XPI showing this "feature" attached.

[userinfotest.xpi.zip](https://github.com/MoonchildProductions/UXP/files/2430230/userinfotest.xpi.zip)

### PR DESCRIPTION
**Pull request by g4jc**
Fixes #795

### CODE REVIEW

File: toolkit/components/startup/moz.build [hunk 1]
```diff
@@ -18,6 +18,7 @@ UNIFIED_SOURCES += [
     'StartupTimeline.cpp',
 ]
 
+#ifdef MOZ_USERINFO
```
Comment (mattatobin) [inline toolkit/components/startup/moz.build]: You don't use preprocessor directives in moz.build files.. You use python in moz.build files... This is all kinds of wrong.

File: toolkit/components/startup/moz.build [hunk 2]
```diff
@@ -31,6 +32,7 @@ else:
     UNIFIED_SOURCES += [
         'nsUserInfoUnix.cpp',
     ]
+#endif
```
Comment (mattatobin) [inline toolkit/components/startup/moz.build]: ditto

File: toolkit/components/startup/public/moz.build [hunk 3]
```diff
@@ -6,7 +6,9 @@
 
 XPIDL_SOURCES += [
     'nsIAppStartup.idl',
+#ifdef MOZ_USERINFO
```
Comment (mattatobin) [inline toolkit/components/startup/public/moz.build]: ditto

File: toolkit/components/startup/public/moz.build [hunk 4]
```diff
@@ -6,7 +6,9 @@
...
     'nsIUserInfo.idl',
+#endif
```
Comment (mattatobin) [inline toolkit/components/startup/public/moz.build]: ditto

File: toolkit/components/startup/moz.build (hunk 1 above)
Comment (wolfbeast) [reply toolkit/components/startup/moz.build]: What you'd want here is something similar to:
`if CONFIG['MOZ_USERINFO']:`
to designate a code block to execute

**Comment by mattatobin** [comment]
r+ but I am still against it.

**Comment by wolfbeast** [comment]
Backed out in f71c04d due to build bustage:

```
17:16.62 nsToolkitCompsModule.cpp
17:16.62 e:/mozdev/UXP/toolkit/components/build/nsToolkitCompsModule.cpp(189): error C2065: 'kNS_USE
RINFO_CID': undeclared identifier
17:16.64
17:16.66 In the directory  /e/mozdev/UXP/obj-pm-x86-unstable/toolkit/components/build
17:16.66 The following command failed to execute properly:
17:16.66 e:/mozdev/UXP/obj-pm-x86-unstable/_virtualenv/Scripts/python.exe -m mozbuild.action.cl d:/V
S2015/VC/BIN/amd64_x86/cl.EXE -FonsToolkitCompsModule.obj -c -Ie:/mozdev/UXP/obj-pm-x86-unstable/dis
t/stl_wrappers -DNDEBUG=1 -DTRIMMED=1 -DSTATIC_EXPORTABLE_JS_API -DMOZ_HAS_MOZGLUE -DMOZILLA_INTERNA
L_API -DIMPL_LIBXUL -Ie:/mozdev/UXP/toolkit/components/build -Ie:/mozdev/UXP/obj-pm-x86-unstable/too
lkit/components/build -Ie:/mozdev/UXP/toolkit/xre -Ie:/mozdev/UXP/toolkit/components/alerts -Ie:/moz
dev/UXP/toolkit/components/downloads -Ie:/mozdev/UXP/toolkit/components/feeds -Ie:/mozdev/UXP/toolki
t/components/find -Ie:/mozdev/UXP/toolkit/components/jsdownloads/src -Ie:/mozdev/UXP/toolkit/compone
nts/perfmonitoring -Ie:/mozdev/UXP/toolkit/components/protobuf -Ie:/mozdev/UXP/toolkit/components/st
artup -Ie:/mozdev/UXP/toolkit/components/statusfilter -Ie:/mozdev/UXP/toolkit/components/typeaheadfi
nd -Ie:/mozdev/UXP/toolkit/components/url-classifier -Ie:/mozdev/UXP/obj-pm-x86-unstable/dist/includ
e -Ie:/mozdev/UXP/obj-pm-x86-unstable/dist/include/nspr -Ie:/mozdev/UXP/obj-pm-x86-unstable/dist/inc
lude/nss -MD -FI e:/mozdev/UXP/obj-pm-x86-unstable/mozilla-config.h -DMOZILLA_CLIENT -TP -nologo -wd
5026 -wd5027 -Zc:sizedDealloc- -Zc:threadSafeInit- -wd4091 -wd4577 -D_HAS_EXCEPTIONS=0 -W3 -Gy -Zc:i
nline -utf-8 -arch:SSE2 -FS -Gw -wd4251 -wd4244 -wd4267 -wd4345 -wd4351 -wd4800 -wd4595 -we4553 -GR-
 -Zi -O2 -Oy- -Fdgenerated.pdb e:/mozdev/UXP/toolkit/components/build/nsToolkitCompsModule.cpp
17:16.68 e:/mozdev/UXP/config/rules.mk:893: recipe for target 'nsToolkitCompsModule.obj' failed
17:16.68 mozmake.EXE[5]: *** [nsToolkitCompsModule.obj] Error 1
17:16.68 e:/mozdev/UXP/config/recurse.mk:71: recipe for target 'toolkit/components/build/target' fai
led
17:16.68 mozmake.EXE[4]: *** [toolkit/components/build/target] Error 2
```

**Comment by wolfbeast** [comment]
Obvious errors in the PR, several instances of `NS_USERINFO` instead of `MOZ_USERINFO`. Was a build even attempted with this code?

**Comment by mattatobin** [comment]
wait.. fuck sakes.. how did he fuck up a patch I wrote for him even though I wasn't supposed to.

**Comment by g4jc** [comment]
I actually do remember having a working build a week or so ago, but I guess my tree was dirty... After clobbering and refreshing the tree, I see the failures are pretty clear. This has turned out to be my most embarrassing ticket for something so seemingly simple thus far. Inserting #ifdef build system code is like walking through land mines, even some of the mingw64 patches were easier.

**Pull request merged**